### DISCUSSION TITLE
How do I set the size of the rpool

### DISCUSSION THREAD

**Discussion by bossey1** [Sithuk/ubuntu-server-zfsbootmenu#53] [Q&A] [answered]
Hi
I just followed this wonderful script:

https://github.com/Sithuk/ubuntu-server-zfsbootmenu

and I can't find anywhere to set or limit the size of the rpool/ it's setting itself on part 3 of my disk and taking up a whole 499GB. the efi is taking only 537 and  and the swap 534. I would like to give rpool 100GB and use the rest for datapool. How do i do this?

Also I have 2 500GB disks and when I set mirror instead of single it results with separate efi partitions, mirrored swap and separate rpools. so of course not surprisingly,  the whole thing refuses to boot. Am I choosing the wrong thing or should I use single and then partition and add the 2nd disk after installation completes with zpool add or zpool attach?

**Comment by Sithuk** [comment, marked as answer]
Hi bossey,

"I can't find anywhere to set or limit the size of the rpool"

The script is setup to create the root pool and data pool on different physical disks. The rpool partition takes up the remainder of the capacity of the disk after the swap and EFI partitions are created. If you wanted to have two pools on a single disk then you would need to create a partition for the other pool and then create the other pool in that partition. The script isn't configured to do that automatically, so you would need to amend the script to achieve the configuration you want.

"Am I choosing the wrong thing or should I use single and then partition and add the 2nd disk after installation completes with zpool add or zpool attach?"

Setting the topology_root variable to mirror and the disks_root variable to 2 should achieve an rpool with a two disk mirror topology. You shouldn't need any manual steps. The separate EFI partitions shouldn't be an issue. rEFInd will auto-select one to boot from, or you can intervene and select which to boot from at the rEFInd menu screen. The contents of both EFI partitions should be identical so it doesn't matter which EFI partition the system boots zfsbootmenu from. There have been other threads about boot issues with particular hardware configurations so the issue may not be the script.

**Reply by bossey1** [reply]
Thank you @Sithuk for your prompt reply.
I will make those amendments and see if it resolves.

**Reply by bossey1** [reply]
> Hi bossey,
> 
> "I can't find anywhere to set or limit the size of the rpool"
> 
> The script is setup to create the root pool and data pool on different physical disks. The rpool partition takes up the remainder of the capacity of the disk after the swap and EFI partitions are created. If you wanted to have two pools on a single disk then you would need to create a partition for the other pool and then create the other pool in that partition. The script isn't configured to do that automatically, so you would need to amend the script to achieve the configuration you want.
> 
> "Am I choosing the wrong thing or should I use single and then partition and add the 2nd disk after installation completes with zpool add or zpool attach?"
> 
> Setting the topology_root variable to mirror and the disks_root variable to 2 should achieve an rpool with a two disk mirror topology. You shouldn't need any manual steps. The separate EFI partitions shouldn't be an issue. rEFInd will auto-select one to boot from, or you can intervene and select which to boot from at the rEFInd menu screen. The contents of both EFI partitions should be identical so it doesn't matter which EFI partition the system boots zfsbootmenu from. There have been other threads about boot issues with particular hardware configurations so the issue may not be the script.

I have now been able to install with 2 EFI parts, mirrored boot and mirrored rpool, by specifying a size of 102400 MiB for the rpool.
It left a free space of 400G but each time I run the script with datapool option it rewrites the whole disk. What edits do I need to make.  I set datapool part to 4 but that makes no difference.

**Reply by Sithuk** [reply]
Could you paste the contents of /tmp/data_creation.sh ? Check for any sensitive information before posting the contents.

**Reply by bossey1** [reply]
Is that after initial or postreboot?

**Reply by Sithuk** [reply]
It should be created as part of running the datapool creation function. The file shouldn't be created until you run the data pool creation function at least once. I'd like to see what commands are being run to create the data pool.

**Reply by bossey1** [reply]
okay here's the disk layout:

`sudo lsblk
NAME    MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0     7:0    0     4K  1 loop /snap/bare/5
loop1     7:1    0  74.2M  1 loop /snap/core22/1439
loop2     7:2    0 505.1M  1 loop /snap/gnome-42-2204/176
loop3     7:3    0 269.3M  1 loop /snap/firefox/4757
loop4     7:4    0  91.7M  1 loop /snap/gtk-common-themes/1535
loop5     7:5    0  38.8M  1 loop /snap/snapd/21759
sda       8:0    0 465.8G  0 disk 
├─sda1    8:1    0   512M  0 part /boot/efi
├─sda2    8:2    0     1G  0 part 
└─sda3    8:3    0   100G  0 part 
sdb       8:16   0 465.8G  0 disk 
├─sdb1    8:17   0   512M  0 part /boot/efi1
├─sdb2    8:18   0     1G  0 part 
│ └─md0   9:0    0     0B  0 md   
└─sdb3    8:19   0   100G  0 part 
sr0      11:0    1  1024M  0 rom  
nvme0n1 259:0    0 465.8G  0 disk 
`
# Starting datapool script

`$ sudo bash ./ubuntu_server_encrypted_root_zfs.sh datapool
Boot environment check passed. Found EFI boot environment.
Mon 19 Aug 15:51:17 BST 2024
               Local time: Mon 2024-08-19 15:51:17 BST
           Universal time: Mon 2024-08-19 14:51:17 UTC
                 RTC time: Mon 2024-08-19 14:51:17
                Time zone: Europe/London (BST, +0100)
System clock synchronized: yes
              NTP service: active
          RTC in local TZ: no
● systemd-timesyncd.service - Network Time Synchronization
     Loaded: loaded (/usr/lib/systemd/system/systemd-timesyncd.service; enabled; preset: enabled)
     Active: active (running) since Mon 2024-08-19 15:51:17 BST; 6ms ago
       Docs: man:systemd-timesyncd.service(8)
   Main PID: 13733 (systemd-timesyn)
     Status: "Connecting to time server 128.138.140.44:123 (128.138.140.44)."
      Tasks: 2 (limit: 18904)
     Memory: 1.4M (peak: 1.8M)
        CPU: 40ms
     CGroup: /system.slice/systemd-timesyncd.service
             └─13733 /usr/lib/systemd/systemd-timesyncd

Aug 19 15:51:17 poma systemd[1]: Starting systemd-timesyncd.service - Network Time Synchronization...
Aug 19 15:51:17 poma systemd[1]: Started systemd-timesyncd.service - Network Time Synchronization.
               Local time: Mon 2024-08-19 15:51:17 BST
           Universal time: Mon 2024-08-19 14:51:17 UTC
                 RTC time: Mon 2024-08-19 14:51:17
                Time zone: Europe/London (BST, +0100)
System clock synchronized: yes
              NTP service: active
          RTC in local TZ: no
Running create data pool on non-root drive. Press Enter to Continue or CTRL+C to abort.
`
***WARNING*** This script could wipe out all your data, or worse! I am not responsible for your decisions. Press Enter to Continue or CTRL+C to abort.

cannot open 'tank1': no such pool
tank1 pre-existance check passed.
Warning: Encryption selected. If the root pool is also encrypted then the root pool keyfile will be used to auto unlock the data pool at boot. Press Enter to Continue or CTRL+C to abort.

Checking script variables for data pool...
User defined topology for data pool: mirror
User defined number of disks in pool: 2
Minimum disk topology check passed for data pool.

Carefully enter the ID of the disk(s) YOU WANT TO DESTROY in the next step to ensure no data is accidentally lost.
The data pool disk topology is mirror with 2 disks.
Please enter Disk ID option for disk 1 of 2 on data pool.
     1	ata-hp_PLDS_DVDRW_DU8AESH_XHDFPGTAU5HP ../../sr0
     2	ata-WDC_WDS500G1R0A-25A7W0_24101S800916 ../../sda
     3	ata-WDC_WDS500G2B0A-00SM16_182232800246 ../../sdb
     4	nvme-eui.e8238ba6bf580002001b448b47a06459 ../../nvme0n1
     5	nvme-WD_Red_SN700_500GB_12345T800948 ../../nvme0n1
     6	nvme-WD_Red_SN700_500GB_12345T800948_1 ../../nvme0n1
     7	scsi-0ATA_WDC_WDS500G1R0A_24101S900718 ../../sda
     8	scsi-0ATA_WDC_WDS500G2B0A-_182232811164 ../../sdb
     9	scsi-1ATA_WDC_WDS500G1R0A-68A4W0_28101S900718 ../../sda
    10	scsi-1ATA_WDC_WDS500G2B0A-00SM50_182232811164 ../../sdb
    11	scsi-35001b448b60db151 ../../sdb
    12	scsi-35001b448c456d446 ../../sda
    13	scsi-SATA_WDC_WDS500G1R0A_28101S900718 ../../sda
    14	scsi-SATA_WDC_WDS500G2B0A-_182232900064 ../../sdb
    15	wwn-0x5001b448b60db151 ../../sdb
    16	wwn-0x5001b448c456d446 ../../sda


Option number 2 selected: ata-WDC_WDS500G1R0A-25A7W0_24101S800916'
Please enter Disk ID option for disk 2 of 2 on data pool.

Select option: 
Option number 3 selected: 'ata-WDC_WDS500G2B0A-00SM16_182232800246'

Clearing partition table on disk ata-WDC_WDS500G1R0A-68A4W0_24101S800718.
Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.

Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
Clearing partition table on disk ata-WDC_WDS500G2B0A-00SM16_182232800246.
Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
Error: Partition(s) 1, 3 on /dev/sda have been written, but we have been unable to inform the kernel of the change, probably because it/they are in use.  As a result, the old partition(s) will remain in use.  You should reboot now before making further changes.
Error: Partition(s) 1, 2, 3 on /dev/sdb have been written, but we have been unable to inform the kernel of the change, probably because it/they are in use.  As a result, the old partition(s) will remain in use.  You should reboot now before making further changes.

I think because of the above error the datapool never gets created so there is no /tmp/data_creation.sh

**Comment by Sithuk** [comment]
What happens when you comment out "clear_partition_table "data"" on line 2074?

So change:
clear_partition_table "data"
to:
#clear_partition_table "data"

**Reply by bossey1** [reply]
Okay just rebooting and will try

**Reply by bossey1** [reply]
> What happens when you comment out "clear_partition_table "data"" on line 2074?
> 
> So change: clear_partition_table "data" to: #clear_partition_table "data"

It just stops. I assume there is a check to see if the whole disk has been cleared.

I've added screenshots of after the datapool script is run.
The only difference i can see is that the partition3 on sdb shows as not mounted and not a part of a mirror. Partition 2 still shows as a raid (mirror) member. It's the same on sda.

![Screenshot from 2024-08-19 21-57-42](https://github.com/user-attachments/assets/db10453e-5943-4d9e-ab21-916fd8404cfe)
![Screenshot from 2024-08-19 21-57-39](https://github.com/user-attachments/assets/0fc14d00-378b-4f43-8be0-6048c6799626)
![Screenshot from 2024-08-19 21-57-25](https://github.com/user-attachments/assets/ae7bf37f-8478-435a-960a-90e533d3c4fc)
![Screenshot from 2024-08-19 21-57-04](https://github.com/user-attachments/assets/8fcc90c4-e971-460f-af95-998c420d77d3)

I also noticed this on reboot as it took a long time to boot.

`2024-08-19T18:15:04.568328+01:00 paloma systemd[1]: Starting systemd-udev-settle.service - Wait for udev To Complete Device Initialization...
2024-08-19T18:15:04.568334+01:00 paloma udevadm[732]: systemd-udev-settle.service is deprecated. Please fix zfs-import-cache.service, zfs-load-module.service not to pull it in.
`

**Reply by Sithuk** [reply]
There isn't a check that the whole disk has been cleared. Was there a /tmp/data_creation.sh file created? 

Uncomment "set -x" on line 9. So change "#set -x" to "set -x". That'll provide more information when the script runs to help debug.

**Reply by bossey1** [reply]
Okay thanks. I'll be out all day today so will do that tonight. Thanks for your patience.

1./
I was playing around and decided to use the single install last night which went without a hitch.  Could it be best to do that then zpool create a datapool on part 4 and then use zpool attach to add mirrors to all the partitions?

2./
Also came across this and after a cursory look, seems to me as if a combimation of this and the above script may achieve want I'm after.  Thoughts?

[https://github.com/peternickol/script.zfsbootmenu.ubuntu/blob/master/zfsbootmenu.sh](url)

**Reply by Sithuk** [reply]
"I was playing around and decided to use the single install last night which went without a hitch. Could it be best to do that then zpool create a datapool on part 4 and then use zpool attach to add mirrors to all the partitions?"

I agree that would probably be the easiest way forward.

I'm not familiar with the script you linked. The quickest way forward is likely to be manually creating the datapool. You can then work out the bugs with modifying this script to your configuration at your leisure once you have a working system.
